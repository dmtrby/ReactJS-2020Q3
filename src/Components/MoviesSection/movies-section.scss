.main {
  width: 100%;
  min-height: 500px;
  background: $black;
}