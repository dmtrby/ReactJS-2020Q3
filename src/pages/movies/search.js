import React, { useEffect } from 'react';
import { useRouter } from 'next/router';
import { useDispatch } from 'react-redux';
import { fetchMovies } from '../../redux/actions';
import Head from 'next/head';

import MainPageHeader from '../../Components/MainPageHeader';
import HeaderSection from '../../Components/HeaderSection';
import FooterSection from '../../Components/FooterSection';
import Main from '../../Components/Main';

// import sprite from './icons/icons-sprite.svg';

function MoviesPage({ noFetch }) {
  const dispatch = useDispatch();

  useEffect(() => {
    if (!noFetch) {
      const router = useRouter();
      const {
        query: { filter, sortBy, search },
      } = router;
      dispatch(fetchMovies(filter, sortBy, search));
    }
  }, [dispatch]);

  return (
    <>
      <Head>
        <title>Movies gallery | Netflix roulette</title>
        <meta name="description" content="Movies gallery page" />
      </Head>
      <HeaderSection>
        <MainPageHeader></MainPageHeader>
      </HeaderSection>
      <Main isEmpty={false} />
      <FooterSection />
    </>
  );
}

MoviesPage.getInitialProps = async ({ store, query }) => {
  const { filter, sortBy, search } = query;
  await store.dispatch(fetchMovies(filter, sortBy, search));
  return {
    noFetch: true,
  };
};

export default MoviesPage;
