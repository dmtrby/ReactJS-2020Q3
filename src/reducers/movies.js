const initialMoviesState = {
  movies: [],
  isLoading: false,
  error: null,
  isDetailsPage: false,
  detailsMovieData: null,
  filter: 'All',
  sort: 'title',
  needToUpdate: true,
};

const movies = (state = initialMoviesState, action) => {
  switch (action.type) {
    case 'FETCH_MOVIES_BEGIN':
      return {
        ...state,
        isLoading: true,
        error: null,
        movies: [],
      };
    case 'FETCH_MOVIES_SUCCESS':
      return {
        ...state,
        isLoading: false,
        movies: action.payload.movies,
        needToUpdate: false,
      };
    case 'ADD_MOVIE_SUCCESS':
      return {
        ...state,
        needToUpdate: true,
      };
    case 'DELETE_MOVIE_SUCCESS':
      return {
        ...state,
        needToUpdate: true,
      };
    case 'DELETE_MOVIE_FAILURE':
      return {
        ...state,
        needToUpdate: true,
      };
    case 'EDIT_MOVIE_SUCCESS':
      return {
        ...state,
        needToUpdate: true,
      };
    case 'CHANGE_FILTER':
      return {
        ...state,
        filter: action.filter,
      };
    case 'CHANGE_SORT':
      return {
        ...state,
        sort: action.sort,
      };
    case 'SET_MAIN_PAGE':
      return {
        ...state,
        isDetailsPage: false,
        detailsMovieData: null,
      };
    case 'SET_DETAILS_PAGE':
      return {
        ...state,
        isDetailsPage: true,
        detailsMovieData: state.movies.find((v) => v.id === action.id),
      };
    default:
      return state;
  }
};

export default movies;
