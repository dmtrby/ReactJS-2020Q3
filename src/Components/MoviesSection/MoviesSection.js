/* eslint-disable react/forbid-prop-types */
import React from 'react';
import { useSelector } from 'react-redux';

import MoviesContainer from '../MoviesContainer';
import DisplayPageStatus from '../DisplayPageStatus';

import './movies-section.scss';

const MoviesSection = () => {
  const { movies, error, isLoading } = useSelector((state) => state);
  return (
    <>
      {movies.length !== 0 && <MoviesContainer movies={movies} />}
      {isLoading && <DisplayPageStatus>Loading...</DisplayPageStatus>}
      {error && <DisplayPageStatus>Error</DisplayPageStatus>}
      {movies.length === 0 && !isLoading && <DisplayPageStatus>No movies Found. Use search tool</DisplayPageStatus>}
    </>
  );
};

export default MoviesSection;
