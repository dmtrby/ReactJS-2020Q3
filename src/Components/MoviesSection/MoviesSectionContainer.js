import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

import { connect} from 'react-redux';
import MoviesSection from './MoviesSection';
import { fetchMovies } from '../../redux/actions';

const MoviesSectionContainer = ({ fetchMoviesFromServer, filter, sortBy, search, ...state }) => {
  useEffect(() => {
    fetchMoviesFromServer(filter, sortBy, search);
  }, [filter, sortBy, search]);

  return <MoviesSection {...state} />;
};

const mapDispatchToProps = (dispatch) => ({
  fetchMoviesFromServer: (filter, sortBy, search) => dispatch(fetchMovies(filter, sortBy, search)),
});

const mapStateToProps = (state) => ({
  movies: state.movies.movies,
  sortBy: state.movies.sortBy,
  search: state.movies.search,
  filter: state.movies.filter,
  isLoading: state.movies.isLoading,
  error: state.movies.error,
});

MoviesSectionContainer.getInitialProps = async ({ query, req }) => {
  await fetchMovies('Horror', 'title', 'd');
  return {};
}

MoviesSectionContainer.propTypes = {
  fetchMoviesFromServer: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(MoviesSectionContainer);
