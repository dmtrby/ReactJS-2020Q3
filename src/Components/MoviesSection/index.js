import MoviesSection from './MoviesSection';

export default MoviesSection;
