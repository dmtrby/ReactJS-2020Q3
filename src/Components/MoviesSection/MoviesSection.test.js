import React from 'react';
import MoviesSection from './MoviesSection';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { initialMoviesState } from '../../redux/reducers/movies';

const middleWares = [thunk];
const mockStore = configureMockStore(middleWares);

const mockedMoviesData = [
  {
    genres: ['Action', 'Adventure', 'Fantasy'],
    id: '1',
    overview: 'overviewText',
    poster_path: 'https://image.tmdb.org/t/p/w500/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg',
    release_date: '2017-10-25',
    runtime: 130,
    title: 'Thor: Ragnarok',
  },
  {
    genres: ['Action', 'Horror', 'Fantasy'],
    id: '2',
    overview: 'overviewText',
    poster_path: 'https://image.tmdb.org/t/p/w500/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg',
    release_date: '2017-10-25',
    runtime: 130,
    title: 'Aviator',
  },
  {
    genres: ['Action', 'Adventure'],
    id: '3',
    overview: 'overviewText',
    poster_path: 'https://image.tmdb.org/t/p/w500/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg',
    release_date: '2017-10-25',
    runtime: 130,
    title: 'ScarFace',
  },
];

test('MoviesSection with movies should matches the Snapshot', () => {
  const store = mockStore(initialMoviesState);
  store.movies = mockedMoviesData;
  const mockedData = {
    isLoading: false,
    movies: mockedMoviesData,
  };
  const component = renderer.create(
    <Provider store={store}>
      <Router>
        <MoviesSection {...mockedData} />
      </Router>
    </Provider>,
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('MoviesSection in loading state should matches the Snapshot', () => {
  const store = mockStore(initialMoviesState);
  store.isLoading = true;
  const mockedData = {
    isLoading: true,
    movies: [],
  };
  const component = renderer.create(
    <Provider store={store}>
      <Router>
        <MoviesSection {...mockedData} />
      </Router>
    </Provider>,
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('MoviesSection in error state should matches the Snapshot', () => {
  const store = mockStore(initialMoviesState);
  store.error = {};
  const mockedData = {
    error: {},
    isLoading: false,
    movies: [],
  };
  const component = renderer.create(
    <Provider store={store}>
      <Router>
        <MoviesSection {...mockedData} />
      </Router>
    </Provider>,
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
