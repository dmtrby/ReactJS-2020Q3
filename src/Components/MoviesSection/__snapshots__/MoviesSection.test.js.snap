// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesSection in error state should matches the Snapshot 1`] = `
<div
  className="page-status container"
>
  <div
    className="row center-xs"
  >
    <div
      className="col-xs-12"
    >
      No movies Found. Use search tool
    </div>
  </div>
</div>
`;

exports[`MoviesSection in loading state should matches the Snapshot 1`] = `
<div
  className="page-status container"
>
  <div
    className="row center-xs"
  >
    <div
      className="col-xs-12"
    >
      No movies Found. Use search tool
    </div>
  </div>
</div>
`;

exports[`MoviesSection with movies should matches the Snapshot 1`] = `
<div
  className="page-status container"
>
  <div
    className="row center-xs"
  >
    <div
      className="col-xs-12"
    >
      No movies Found. Use search tool
    </div>
  </div>
</div>
`;
